[#592] [도서 검색] 검색 키워드 및 결과 유지 기능 구현

harryk-im opened the pull request:
# 구현 내용
- 검색 키워드 & 결과 유지 기능 추가
- 검색창에 type="search" 추가, x-button 스타일링
- [useQueryParams](https://github.com/prgrms-web-devcourse/Team-Gaerval-Dadok-FE/pull/593/commits/7baecebd41deb26bc86662ca16aea7dd29c4bd59) 커스텀 훅 작성

# 스크린샷
https://github.com/prgrms-web-devcourse/Team-Gaerval-Dadok-FE/assets/73218463/cfe31783-d42e-4321-a45c-7d0a6a75d48d


# pr 포인트
쿼리 파라미터와 react-hook-form관련 코드들의 관심사를 최대한 섞이지 않도록 작성해보았어요.
자세한 내용은 코멘트로 남겨두었어요!


# 관련 이슈

- Close #592

harryk-im committed:
feat: useQueryParams 커스텀 훅 작성

harryk-im committed:
feat: 검색 키워드 유지 기능 추가

harryk-im committed:
fix: Skeleton 스타일 수정

harryk-im committed:
fix: global.css에 warning이 나타나던 문제 해결

harryk-im committed:
feat: 검색 결과 유지 기능 추가

harryk-im committed:
chore: 변수명 수정

harryk-im committed:
chore: 주석 추가

harryk-im committed:
feat: 도서 검색 Input에 검색어 초기화 기능 추가

hanyugeon commented on src/hooks/useQueryParams.ts:
```diff
@@ -0,0 +1,65 @@
+
+type RouteOptions = 'push' | 'replace';
+
+const useQueryParams = () => {
+  const router = useRouter();
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+
```
comment;

URL의 쿼리파라미터를 관리할 수 있도록 만든 `useQueryParams` 커스텀 훅이에요
return값에는 `queryParams`, `getQueryParam`, `setQueryParams`, `removeQueryParam` 이 존재합니다

`queryParams`: URL의 쿼리파라미터 전체를 string으로 반환
`getQueryParam`: queryKey를 인자로 받아 queryValue를 return
`setQueryParams`: queryKey와 queryValue를 인자로 받아 해당 query를 URL에 추가하고 router.push() 혹은 replace() 동작
`removeQueryParam`: queryKey를 인자로 받아 해당 query를 URL에서 삭제하고 router.push() 혹은 replace() 동작

gxxrxn commented on src/hooks/useQueryParams.ts:
```diff
@@ -0,0 +1,65 @@
+    searchParams.toString()
+  );
+
+  const getQueryParam = (queryKey: string) => {
+    const queryParam = searchParams.get(queryKey);
+
+    return queryParam;
+  };
```
p1;

각 함수들이 effect의 의존성으로 들어가고 있는만큼 각 함수를 useCallback으로 감싸서 effect 함수 실행을 최소화하면 좋을 것 같아요! [(참고)](https://ko.react.dev/reference/react/useCallback#should-you-add-usecallback-everywhere)

hanyugeon replied:
@gxxrxn 

[ba3009e](https://github.com/prgrms-web-devcourse/Team-Gaerval-Dadok-FE/pull/593/commits/ba3009e99c8a701c0d83664ebe54959ea22b3d91) 반영하였습니다
useCallback을 활용하니 useEffect의 발생 빈도가 줄어들었네요! 감사합니다 🙇

gxxrxn commented on src/hooks/useQueryParams.ts:
```diff
@@ -0,0 +1,65 @@
+
+  const removeQueryParam = (queryKey: string, option?: RouteOptions) => {
+    const prevParams = new URLSearchParams(searchParams.toString());
+    if (!prevParams.get(queryKey)) return;
```
p3;

boolean을 반환하는 `URLSearchParams.has` 메소드를 사용해도 좋을 것 같아요~

hanyugeon replied:
@gxxrxn 

has라는 메소드도 있었군요 🤔 
반영하였습니다! [ba3009e](https://github.com/prgrms-web-devcourse/Team-Gaerval-Dadok-FE/pull/593/commits/ba3009e99c8a701c0d83664ebe54959ea22b3d91)

gxxrxn commented on src/hooks/useQueryParams.ts:
```diff
@@ -0,0 +1,65 @@
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+
+  const [queryParams, _setQueryParams] = useState<string>(
+    searchParams.toString()
+  );
```
p3;

상태에 따른 UI 변화가 필요한게 아니기 때문에 state로 쿼리 파라미터를 관리하지 않아도 될 것 같다는 생각이 들었어요! 해당 hook에서 쿼리파라미터를 따로 관리하지 않고, useSearchParams를 통해 바로 사용하는 형태로 구현해볼 수 있지 않을까요?! 🧐

hanyugeon replied:
@gxxrxn 

좋아요! 불필요하게 사용된 state를 줄일 수 있어 좋은 것 같아요!

gxxrxn reviewed (commented):
수고많으셨습니다!!🙌🏻 검색 결과 유지되니까 너무 좋은 것 같아요! 👍🏻 코멘트 몇개 확인해주시면 감사하겠습니당~🙇🏻‍♀️

harryk-im committed:
fix: unable to verify the first certificate 인증서를 인식못하던 문제 해결

harryk-im committed:
refactor: useQueryParams 개선

- 불필요한 state 제거
- useCallback 훅 적용
- 잘못 포함된 '/' 제거
- removeQueryParam의 earlyReturn 문 개선

harryk-im committed:
refactor: queryParams 개선 (state 사용x)

harryk-im committed:
fix: 검색창 search 아이콘 LayoutShift 현상 제거

harryk-im committed:
fix: setQueryParams 개선

gxxrxn reviewed (approved):
👍🏻👍🏻
